torch/csrc/jit/passes/graph_fuser.cpp

```
@@ -249,18 +329,53 @@ struct GraphFuser {
     return isFusableOnlyAsExitNode(node);
   }
 
+  // unknown (u) any (a) cpu (c) cuda (g) compatibility:
+  // x u a c g   y = yes
+  // u . . . .   . = no
+  // a . y y y
+  // c . y y .
+  // g . y . y
+  bool compatibleDevices(Node * consumer, Value * producer) {
+    auto consumer_device = getDevice(consumer);
+    auto producer_device = getDevice(producer->node());
+
+    if (consumer_device.type() == DeviceType::Unknown ||
+        producer_device.type() == DeviceType::Unknown) {
+      return false;
+    }
+
+    if (consumer_device.type() == DeviceType::CUDA &&
+        producer_device.type() == DeviceType::CPU) {
+      return false;
+    } else if (producer_device.type() == DeviceType::CUDA &&
+               consumer_device.type() == DeviceType::CPU) {
+      return false;
+    } else if (producer_device.type() == DeviceType::AnyDevice &&
+               consumer_device.type() == DeviceType::AnyDevice) {
+      // XXX: This case means we're fusing operations on non-constant numbers.
+      // The graph fuser doesn't support this at the moment (#9940).
+      JIT_ASSERT(false);
```

Why do we have a hard failure in here instead of returning false?